### PR TITLE
Gh 587

### PR DESCRIPTION
**Pull request by philbrookes**
- add TLS Policy integration tests
- Minor tweaks to tests to stop flaky failures and fix issue with wildcard listeners in utility package.
### verification
- see PR tests pass

### CODE REVIEW

**Comment by philbrookes** [comment]
related to integration sections of: https://github.com/Kuadrant/multicluster-gateway-controller/issues/587

**Review by laurafitzgerald — COMMENTED** [review]
One question inline.

File: test/integration/tlspolicy_controller_test.go [hunk 1]
```diff
@@ -336,8 +373,57 @@ var _ = Describe("TLSPolicy", Ordered, func() {
...
+			It("should delete tls certificate when listener is removed", func() {
+				//confirm all expected certificates are present
+				Eventually(func() error {
+					certificateList := &certmanv1.CertificateList{}
+					Expect(k8sClient.List(ctx, certificateList, &client.ListOptions{Namespace: testNamespace})).To(BeNil())
+					if len(certificateList.Items) != 3 {
+						return fmt.Errorf("expected 3 certificates, found: %v", len(certificateList.Items))
+					}
+					return nil
+				})
```
[the remark this answers is not in the document]
Comment (philbrookes) [reply test/integration/tlspolicy_controller_test.go:385]: 🤦‍♂️ Thanks @laurafitzgerald I'll address these now.

File: test/integration/tlspolicy_controller_test.go [hunk 2]
```diff
@@ -80,13 +81,49 @@ var _ = Describe("TLSPolicy", Ordered, func() {
...
+					if freshGW.Status.Conditions == nil {
+						return nil
+					}
+					for _, condition := range freshGW.Status.Conditions {
+						if condition.Type == string(gatewayv1beta1.GatewayConditionProgrammed) {
+							if strings.ToLower(string(condition.Status)) == "true" {
+								return fmt.Errorf("expected programmed status false, got true")
+							}
+						}
+					}
```
Comment (mikenairn) [inline test/integration/tlspolicy_controller_test.go:120]: Why do you get a gateway and update `gateway`, but check the conditions on  `freshGW`. What is this test testing?

File: test/integration/dnspolicy_controller_test.go [hunk 3]
```diff
@@ -1329,10 +1330,9 @@ var _ = Describe("DNSPolicy", Ordered, func() {
 					Protocol: gatewayv1beta1.HTTPProtocolType,
 				}
 
-				Eventually(func() error {
-					gateway.Spec.Listeners = append(gateway.Spec.Listeners, otherListener)
-					return k8sClient.Update(ctx, gateway)
-				}).Should(BeNil())
+				patch = client.MergeFrom(gateway.DeepCopy())
+				gateway.Spec.Listeners = append(gateway.Spec.Listeners, otherListener)
+				Expect(k8sClient.Patch(ctx, gateway, patch)).To(BeNil())
```
[the remark this answers is not in the document]
Comment (philbrookes) [reply test/integration/dnspolicy_controller_test.go:1335]: yes, I'd expect this change to resolve that.

File: test/integration/tlspolicy_controller_test.go (hunk 2 above)
Comment (philbrookes) [reply test/integration/tlspolicy_controller_test.go:120]: bah, this is wrong. I'll fix it now thanks.

File: test/integration/tlspolicy_controller_test.go [hunk 4]
```diff
@@ -322,11 +359,11 @@ var _ = Describe("TLSPolicy", Ordered, func() {
 					certList := &certmanv1.CertificateList{}
 					err := k8sClient.List(ctx, certList, &client.ListOptions{Namespace: testNamespace})
 					Expect(err).ToNot(HaveOccurred())
-					if len(certList.Items) != 2 {
-						return fmt.Errorf("expected CertificateList to be 2")
+					if len(certList.Items) != 3 {
+						return fmt.Errorf("expected CertificateList to be 3")
 					}
```
Comment (mikenairn) [inline test/integration/tlspolicy_controller_test.go:364]: This was originally intended to test multiple listeners, but intentionally had 2 with the same name to check it only created 2 even though there was 3 listeners. 

We probably in reality need more test cases that describe each case better and actually check the created certificates have the correct DNSNames for the expected listeners.

File: test/integration/tlspolicy_controller_test.go [hunk 5]
```diff
@@ -335,9 +372,62 @@ var _ = Describe("TLSPolicy", Ordered, func() {
 				cert2 := &certmanv1.Certificate{}
 				err = k8sClient.Get(ctx, client.ObjectKey{Name: "test2-tls-secret", Namespace: testNamespace}, cert2)
 				Expect(err).ToNot(HaveOccurred())
```
Comment (mikenairn) [inline test/integration/tlspolicy_controller_test.go:374]: Here, we should check the certs have the expected DNSNames:

cert1.Spec.DNsNames = ["test1.example.com"]
cert2.Spec.DNsNames = ["test2.example.com", "test2.example.com"]

File: test/integration/tlspolicy_controller_test.go (hunk 2 above)
Comment (philbrookes) [reply test/integration/tlspolicy_controller_test.go:120]: sorted now.

File: test/integration/tlspolicy_controller_test.go (hunk 4 above)
Comment (philbrookes) [reply test/integration/tlspolicy_controller_test.go:364]: ok, I'll put this one back how it was - I assumed it was a typo - and make a new context for my tests.

File: test/integration/tlspolicy_controller_test.go (hunk 5 above)
Comment (philbrookes) [reply test/integration/tlspolicy_controller_test.go:419]: done.

File: test/integration/tlspolicy_controller_test.go [hunk 6]
```diff
@@ -326,18 +362,121 @@ var _ = Describe("TLSPolicy", Ordered, func() {
...
 					}
 					return nil
-				}, time.Second*10, time.Second).Should(BeNil())
+				}, time.Second*30, time.Second).Should(BeNil())
 
 				cert1 := &certmanv1.Certificate{}
 				err := k8sClient.Get(ctx, client.ObjectKey{Name: "test-tls-secret", Namespace: testNamespace}, cert1)
 				Expect(err).ToNot(HaveOccurred())
 
+				Expect(cert1.Spec.DNSNames).To(BeEquivalentTo([]string{"test1.example.com", "test2.example.com"}))
```
Comment (mikenairn) [inline test/integration/tlspolicy_controller_test.go:371]: Will `BeEquivalentTo` take into account the ordering of the elements? `ConsistOf` is another matcher that checks elements exist and only those elements in an array regardless of order.
```
ConsistOf("test1.example.com", "test2.example.com")
```

**Comment by mikenairn** [comment]
Question about one of the gomega matchers being used, but other than that, changes look good.

File: test/integration/tlspolicy_controller_test.go (hunk 6 above)
Comment (philbrookes) [reply test/integration/tlspolicy_controller_test.go:371]: I had assumed it would based on the name, but I'm not sure, I'll update it to `ConsistOf`

**Pull request merged**